pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go

```diff
@@ -0,0 +1,298 @@
+		syncutil.Mutex
+
+		// terminalErr set when a terminal error occurs.
+		// Subsequent calls to startMuxRangeFeed will return this error.
+		terminalErr error
+
+		// Each call to start new range feed gets a unique ID which is echoed back
+		// by MuxRangeFeed rpc.  This is done as a safety mechanism to make sure
+		// that we always send the event to the correct consumer -- even if the
+		// range feed is terminated and re-established rapidly.
+		nextStreamID int64
+
+		// muxClient contains a nodeID->MuxRangeFeedClient.
+		muxClients map[roachpb.NodeID]*muxClientState
+		// producers maps streamID to the event producer -- data sent back to the
+		// consumer of range feed events.
+		producers map[int64]*channelRangeFeedEventProducer
+	}
+}
+
+type muxClientState struct {
+	client  roachpb.Internal_MuxRangeFeedClient
+	streams util.FastIntSet
+	cancel  context.CancelFunc
+}
+
+func newRangefeedMuxer(g ctxgroup.Group) *rangefeedMuxer {
+	m := &rangefeedMuxer{
+		eventCh: make(chan *roachpb.MuxRangeFeedEvent),
+		g:       g,
+	}
+	m.mu.muxClients = make(map[roachpb.NodeID]*muxClientState)
+	m.mu.producers = make(map[int64]*channelRangeFeedEventProducer)
+
+	m.g.GoCtx(m.demux)
+
+	return m
+}
+
+type eventOrError struct {
+	event *roachpb.RangeFeedEvent
+	err   error
+}
+
+// channelRangeFeedEventProducer is a rangeFeedEventProducer which receives
+// events on input channel, and returns events when Recv is called.
+type channelRangeFeedEventProducer struct {
+	ctx   context.Context
+	errCh chan error // Signalled to propagate terminal error the consumer.
+	input chan eventOrError
+}
+
+// Recv implements rangeFeedEventProducer interface.
+func (c *channelRangeFeedEventProducer) Recv() (*roachpb.RangeFeedEvent, error) {
+	select {
+	case <-c.ctx.Done():
+		return nil, c.ctx.Err()
+	case err := <-c.errCh:
+		return nil, err
+	case e := <-c.input:
+		return e.event, e.err
+	}
+}
+
+var _ roachpb.RangeFeedEventProducer = (*channelRangeFeedEventProducer)(nil)
+
+// startMuxRangeFeed begins the execution of rangefeed for the specified
+// RangeFeedRequest.
+func (m *rangefeedMuxer) startMuxRangeFeed(
+	ctx context.Context, client rpc.RestrictedInternalClient, req *roachpb.RangeFeedRequest,
+) (_ roachpb.RangeFeedEventProducer, cleanup func(), _ error) {
+	producer, rpcClient, streamID, cleanup, err := m.connect(ctx, client, req.Replica.NodeID)
+	if err != nil {
+		return nil, cleanup, err
+	}
+	req.StreamID = streamID
+	if err := rpcClient.Send(req); err != nil {
+		return nil, cleanup, err
+	}
+	return producer, cleanup, nil
+}
+
+// connect establishes MuxRangeFeed connection for the specified node. returns
+// event producer, RPC client to send requests on, the streamID which should be
+// used when sending request, and a cleanup function. Cleanup function never
+// nil, and must always be invoked, even if error is returned.
+func (m *rangefeedMuxer) connect(
+	ctx context.Context, client rpc.RestrictedInternalClient, nodeID roachpb.NodeID,
```

Comment:
It is a little odd that `client` is passed in here. In the common case, it is never used because we use a cached `MuxRangeFeed`, which was created off a different client.
I think this is justifiable as it's an artifact of making the muxing change non-invasively, without changing a lot of DistSender code. But we should write down what's going on here and why it is that way, and point out in the methods that are affected that the passed-in client isn't always used.

Reply:
I think you got this 100% right -- the initial versions were very invasive; and this file exists to make a non-invasive
change (that can also turn off the use of mux range feed) -- added comment in startMuxRangeFeed.